hathor/indexes/rocksdb_timestamp_index.py

```diff
@@ -0,0 +1,151 @@
+
+    def _to_key(self, timestamp: int, tx_hash: Optional[bytes] = None) -> bytes:
+        """Make a key for a timestamp and optionally tx_hash, the key represents the membership itself."""
+        import struct
+        key = bytearray()
+        key.extend(struct.pack('>I', timestamp))
+        if tx_hash is None:
+            assert len(key) == 4
+            return bytes(key)
+        assert len(tx_hash) == 32
+        key.extend(tx_hash)
+        assert len(key) == 4 + 32
+        return bytes(key)
+
+    def _from_key(self, key: bytes) -> Tuple[int, bytes]:
+        """Parse a key on the column-family."""
+        import struct
+        assert len(key) == 4 + 32
+        timestamp: int
+        (timestamp,) = struct.unpack('>I', key[:4])
+        tx_hash = key[4:]
+        assert len(tx_hash) == 32
+        return timestamp, tx_hash
+
+    def add_tx(self, tx: BaseTransaction) -> bool:
+        key = self._to_key(tx.timestamp, tx.hash)
+        is_new = self._db.get((self._cf, key)) is None
+        if is_new:
+            self._db.put((self._cf, key), b'')
+        return is_new
+
+    def del_tx(self, tx: BaseTransaction) -> None:
+        key = self._to_key(tx.timestamp, tx.hash)
+        self._db.delete((self._cf, key))
+
+    def _iter(self, from_timestamp: Optional[int] = None, from_tx: Optional[bytes] = None,
+              *, reverse: bool = False) -> Iterator[Tuple[int, bytes]]:
+        """ Iterate over transactions optionally starting from a timestamp/hash, by default from oldest to newest.
+        """
+        if from_timestamp is None and from_tx is not None:
+            raise ValueError('from_tx needs from_timestamp, but it is None')
+        it = self._db.iterkeys(self._cf)
+        if reverse:
+            it = reversed(it)
+            if from_timestamp is None:
+                self.log.debug('seek to last')
+                it.seek_to_last()
+            else:
+                # when reversed we increment the key by 1, which effectively goes to the end of a prefix
+                self.log.debug('seek to', tx=from_tx)
+                it.seek_for_prev(incr_key(self._to_key(from_timestamp, from_tx)))
+        else:
+            if from_timestamp is None:
+                self.log.debug('seek to first')
+                it.seek_to_first()
+            else:
+                self.log.debug('seek to', tx=from_tx)
+                it.seek(self._to_key(from_timestamp, from_tx))
+        first = True
+        for _, key in it:
+            timestamp, tx_hash = self._from_key(key)
+            self.log.debug('seek found', tx=tx_hash.hex())
+            # XXX: this is made such that we wouldn't skip the first element blindly, only if it is the given from_tx
+            if first:
+                first = False
+                # from_tx transaction is never yielded
+                if from_timestamp is not None and from_timestamp == timestamp and from_tx == tx_hash:
+                    continue
```

Comment:
That's right. But we can't blindly always skip the first element, because if it isn't on the index the seek would still work just fine, but go to the next closest key. While this doesn't happen from "internal" use, this could happen from API calls that look up this index.

Reply:
Got it. I think we should explain it in a comment. I had to think for a while to understand this behavior.

Reply:
Done.